casper/src/test/scala/coop/rchain/casper/HashSetCasperTest.scala

```diff
@@ -1057,6 +1057,101 @@ class HashSetCasperTest extends FlatSpec with Matchers {
     } yield result
   }
 
+  it should "ask peers for blocks it is missing and add them" ignore effectTest {
+    val deployDatasFs = Vector(
+      "@2!(2)",
+      "@1!(1)"
+    ).zipWithIndex
+      .map(
+        d =>
+          () =>
+            ProtoUtil.sourceDeploy(d._1, System.currentTimeMillis() + d._2, accounting.MAX_VALUE)
+      )
+    def deploy(node: HashSetCasperTestNode[Effect], dd: DeployData): Effect[BlockMessage] =
+      for {
+        createBlockResult1    <- node.casperEff.deploy(dd) *> node.casperEff.createBlock
+        Created(signedBlock1) = createBlockResult1
+
+        _ <- node.casperEff.addBlock(signedBlock1, ignoreDoppelgangerCheck[Effect])
+      } yield signedBlock1
+
+    def stepSplit(nodes: Seq[HashSetCasperTestNode[Effect]]) = {
+      for {
+        _ <- deploy(nodes(0), deployDatasFs(0).apply())
+        _ <- deploy(nodes(1), deployDatasFs(1).apply())
+
+        _ <- nodes(0).receive()
+        _ <- nodes(1).receive()
+        _ <- nodes(2).transportLayerEff.clear(nodes(2).local) //nodes(2) misses this block
+      } yield ()
+    }
+
+    def stepSingle(nodes: Seq[HashSetCasperTestNode[Effect]]) = {
+      for {
+        _ <- deploy(nodes(0), deployDatasFs(0).apply())
+
+        _ <- nodes(0).receive()
+        _ <- nodes(1).receive()
+        _ <- nodes(2).transportLayerEff.clear(nodes(2).local) //nodes(2) misses this block
+      } yield ()
+    }
+
+    def propagate(nodes: Seq[HashSetCasperTestNode[Effect]]) = {
+      for {
+        _ <- nodes(0).receive()
+        _ <- nodes(1).receive()
+        _ <- nodes(2).receive() //nodes(2) gets the 5th block and starts requesting the dag
```

Comment:
This tests uses a mocked version of the transport layer. When you call `receive` on a node, it takes things off its mocked network queue and then processes them. If you want to simulate the process of node A asking for a block from node B and having node B send it back, you have to do something like

```
// node A sends block request
nodeB.receive() // node B receives block request and sends the block
nodeA.receive() // node A receives block
```

Reply:
that's why it's done 20 times - to let the network do its thing.

Reply:
I see; so in theory you only need to do like 5~6 times but you just do it 20 times to be sure?

Reply:
How does nodes(2) get the 5th block?

Reply:
> How does nodes(2) get the 5th block?

Nvm ignore

Reply:
in theory yes BUT:
I've observed a funny "self healing" mechanism - if the network layer holds enough requested blocks it will make the catch up mechanism look like it works (it processes stale requested blocks). This way I'm making sure all the blocks are processed.